Support serving images under s3 subdirectories, Fix to make /fit-in/ work; Fix for VipsJpeg: Invalid SOS error plus several other critical fixes.

rpong opened the pull request:
This PR includes fixes for the following v4 issues:

- S3 subdirectories not working.
- /fit-in/ url directive not working.
- Exif orientation not honoured/followed.
- `VipsJpeg: Invalid SOS parameters for sequential JPEG` errors. This is observable with images taken with Samsung cameras, causing the api/cloudfront endpoint to output a blank json `{}` instead of the image.
- Wrong resizing if the path/filename contains an 'x' with surrounding digits (eg: /image8x8.jpg)

Sample paths in which this was tested on aside from the unit tests:

`/100x100/directory1/directory2/image.jpg` 
`/fit-in/100x100/directory1/directory2/image.jpg` 
`/fit-in/100x100/filters:someFilters/directory1/directory2/image.jpg` 
`/filters:someFilters/directory1/directory2/image.jpg` 
`/100x100/filters:someFilters/directory1/directory2/image_8x8x8x12.jpg`

You may start using this PR/Fork with this Exported Function zip: https://github.com/innovationlove/serverless-image-handler/releases/tag/4.0.1-ilove

Just replace the code that your Serverless Image Handler Function is using with the package above. USE AT YOUR OWN RISK.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

beyerz commented:
I have manually added this solution to our lambda-function and can confirm that it performs as defined.
@rpong You are a life saver!

denys1yermakov commented:
@rpong, images without resizing still doesn't work, example:
/directory1/directory2/image.jpg
If you can fix it in that pull request will be super cool, thanks!

codyhawke commented:
Been trying to get this going - everything seems to work fine, except it's not resizing the image when in a subdirectory - example url:
https://xxxxxxx.cloudfront.net/400x0/products/img/product01.jpg

rpong commented:
> @rpong, images without resizing still doesn't work, example:
> /directory1/directory2/image.jpg
> If you can fix it in that pull request will be super helpfull, thanks!

sorry, my bad. this should be fixed now.

rpong commented:
> Been trying to get this going - everything seems to work fine, except it's not resizing the image when in a subdirectory - example url:
> https://xxxxxxx.cloudfront.net/400x0/products/img/product01.jpg

can you attach the actual photo and actual filename? we discovered an anomaly with filenames containing an 'x' and digits arounds it causing improper resizing, we will be applying the fix for the PR too in a bit, just want to make sure about your use case and perhaps cover it with our fix too. thanks man..

gendigbadig commented:
> > Been trying to get this going - everything seems to work fine, except it's not resizing the image when in a subdirectory - example url:
> > https://xxxxxxx.cloudfront.net/400x0/products/img/product01.jpg
> 
> can you attach the actual photo and actual filename? we discovered an anomaly with filenames containing an 'x' and digits around it causing improper resizing, we will be applying the fix for the PR too in a bit, just want to make sure about your use case and perhaps cover it with our fix later today. thanks man..

in my use case, some of key path contains 'x' characters, and i discovered dimension path checking on thumbor-mapping.js is using only `String.prototype.includes` method, so i change those line to using regex match
Line 39
```
           else if (edit.match(/^\d+x\d+$/)) { // change this line
                this.edits.resize = {};
                const dims = edit.split('x');
                this.edits.resize.width = Number(dims[0]);
                this.edits.resize.height = Number(dims[1]);
            } 
```

Please add it to your PR, thank you :)

denys1yermakov commented:
> > > Been trying to get this going - everything seems to work fine, except it's not resizing the image when in a subdirectory - example url:
> > > https://xxxxxxx.cloudfront.net/400x0/products/img/product01.jpg
> > 
> > 
> > can you attach the actual photo and actual filename? we discovered an anomaly with filenames containing an 'x' and digits around it causing improper resizing, we will be applying the fix for the PR too in a bit, just want to make sure about your use case and perhaps cover it with our fix later today. thanks man..
> 
> in my use case, some of key path contains 'x' characters, and i discovered dimension path checking on `thumbor-mapping.js` is using only `String.prototype.includes` method, so i change those line to using regex match
> 
> Line 39
> 
> ```
>            else if (edit.match(/^\d+x\d+$/)) { // change this line
>                 this.edits.resize = {};
>                 const dims = edit.split('x');
>                 this.edits.resize.width = Number(dims[0]);
>                 this.edits.resize.height = Number(dims[1]);
>             } 
> ```
> 
> Please add it to your PR, thank you :)

that work, thanks !

gendigbadig commented:
also, in previous version (thumbor), i have some use case that unfortunately unsupported on current version: 
1. using key path without file extension
Example: https://xxxx.cloudfront.net/64x64/images/svara-radio-logo/download/58e5ecc7cab82dc1700c0043
My workaround is remove path extension checking on `parseRequestType` method on `image-request.js`
```javascript
parseRequestType(event) {
    const path = event["path"];
    // ----
    const matchDefault = new RegExp(
      /^(\/?)([0-9a-zA-Z+\/]{4})*(([0-9a-zA-Z+\/]{2}==)|([0-9a-zA-Z+\/]{3}=))?$/
    );
    const matchThumbor = new RegExp(
      /^(\/?)((fit-in)?|(filters:.+\(.?\))?|(unsafe)?).*$/  // Remove extension on this regex
    ); 
    const matchCustom = new RegExp(/(\/?)(.*)(jpg|png|webp|tiff|jpeg)/);
    const definedEnvironmentVariables =
      process.env.REWRITE_MATCH_PATTERN !== "" &&
      process.env.REWRITE_SUBSTITUTION !== "" &&
      process.env.REWRITE_MATCH_PATTERN !== undefined &&
      process.env.REWRITE_SUBSTITUTION !== undefined;
    // ----
    if (matchDefault.test(path)) {
      // use sharp
      return "Default";
    } else if (matchCustom.test(path) && definedEnvironmentVariables) {
      // use rewrite function then thumbor mappings
      return "Custom";
    } else if (matchThumbor.test(path)) {
      // use thumbor mappings
      return "Thumbor";
    } else {
      throw {
        status: 400,
        code: "RequestTypeError",
        message:
          "The type of request you are making could not be processed. Please ensure that your original image is of a supported file type (jpg, png, tiff, webp) and that your image request is provided in the correct syntax. Refer to the documentation for additional guidance on forming image requests."
      };
    }
  }

```

2. using `smart` path
For now, my workaround is only omit the `smart` keyword on `parseImageKey` method at `image-request.js`
```javascript
parseImageKey(event, requestType) {
    if (requestType === "Default") {
      // Decode the image request and return the image key
      const decoded = this.decodeRequest(event);
      return decoded.key;
    } else if (requestType === "Thumbor" || requestType === "Custom") {
      // Parse the key from the end of the path
      const pathParts = event["path"].split("/");

      for (let i = 0; i < pathParts.length; i++) {
        if (
          pathParts[i] !== "" &&
          pathParts[i] !== "fit-in" &&
          pathParts[i].match(/^\d+x\d+$/) == null &&
          pathParts[i].match(/^filters:/) == null &&
          pathParts[i].match(/^smart$/) == null  // add this line
        ) {
          const parts = pathParts.slice(i);
          return parts.join("/");
        }
      }
    }

    // Return an error for all other conditions
    throw {
      status: 400,
      code: "ImageEdits::CannotFindImage",
      message:
        "The image you specified could not be found. Please check your request syntax as well as the bucket you specified to ensure it exists."
    };
  }
```

3. using filters fill with same usage as thumbor, ex:`filters:fill(white,1)`
Currently my workaround on this problem is adding some hardcoded line to convert filter value to color object that accepted by sharpjs, ex: `'white,1'` to `{r: 255, g:255, b: 255, alpha: 0}` 

maybe you have more clean workaround for this problem
thank you

rpong commented:
> > > Been trying to get this going - everything seems to work fine, except it's not resizing the image when in a subdirectory - example url:
> > > https://xxxxxxx.cloudfront.net/400x0/products/img/product01.jpg
> > 
> > 
> > can you attach the actual photo and actual filename? we discovered an anomaly with filenames containing an 'x' and digits around it causing improper resizing, we will be applying the fix for the PR too in a bit, just want to make sure about your use case and perhaps cover it with our fix later today. thanks man..
> 
> in my use case, some of key path contains 'x' characters, and i discovered dimension path checking on `thumbor-mapping.js` is using only `String.prototype.includes` method, so i change those line to using regex match
> 
> Line 39
> 
> ```
>            else if (edit.match(/^\d+x\d+$/)) { // change this line
>                 this.edits.resize = {};
>                 const dims = edit.split('x');
>                 this.edits.resize.width = Number(dims[0]);
>                 this.edits.resize.height = Number(dims[1]);
>             } 
> ```
> 
> Please add it to your PR, thank you :)

Nice! thank you for this, I added this to the PR but moved it outside the foreach loop so it is processed/looked up only once. Just updated the PR as well so the tests pass and several other critical things that required us to fix before we launched our beta app this morning.

codyhawke commented:
> > Been trying to get this going - everything seems to work fine, except it's not resizing the image when in a subdirectory - example url:
> > https://xxxxxxx.cloudfront.net/400x0/products/img/product01.jpg
> 
> can you attach the actual photo and actual filename? we discovered an anomaly with filenames containing an 'x' and digits around it causing improper resizing, we will be applying the fix for the PR too in a bit, just want to make sure about your use case and perhaps cover it with our fix later today. thanks man..
/600x400/products/img/denial-krylon-8inchdeck-1xrun-00.jpg

codyhawke commented:
Still not working for some of my URLS such as the following:
https://xxxxxxxxx.cloudfront.net/400x400/products/img/ricky-powell-beastie-boys-18x13-1xrun-01.jpg

rpong commented:
> Still not working for some of my URLS such as the following:
> https://xxxxxxxxx.cloudfront.net/400x400/products/img/ricky-powell-beastie-boys-18x13-1xrun-01.jpg

thanks for that use case, just pushed a change to the PR which should take care of it, strange that the unit-tests missed this.

adam91holt commented:
Hi @rpong 

Thanks for all your hard work. Another issue I just spotted is when you form a URL like so...

https://d1h4lz25tmeef3.cloudfront.net/fit-in/400x/clients/demo/images/flip/ichts176-navaksha-original-imaeceu7sqmcekt8.jpeg

Specifying "400x" with the old Thumbor used to give you the best fit for 400 and figure it out itself. However, doing that now will give you "NoSuchKey". If you use 400x400 for example then it works fine.

https://d1h4lz25tmeef3.cloudfront.net/fit-in/400x400/clients/demo/images/flip/ichts176-navaksha-original-imaeceu7sqmcekt8.jpeg

rpong commented:
> https://d1h4lz25tmeef3.cloudfront.net/fit-in/400x400/clients/demo/images/flip/ichts176-navaksha-original-imaeceu7sqmcekt8.jpeg

Hi @adam91holt , noted on this, for now, if this is an actual production use case for you, what we do to circumvent around this is to specify zero and a fit-in.

https://d1h4lz25tmeef3.cloudfront.net/fit-in/400x0/clients/demo/images/flip/ichts176-navaksha-original-imaeceu7sqmcekt8.jpeg

I did not realize they support 400x before, such dimension convention also wasn't in the unit tests so we did not cover that.

ahmetcetin39 commented on source/image-handler/thumbor-mapping.js:
```diff
@@ -29,20 +29,31 @@ class ThumborMapping {
         const edits = this.path.split('/');
         const filetype = (this.path.split('.'))[(this.path.split('.')).length - 1];
+
+        //Process the Dimenions
```
Dimensions?

ahmetcetin39 commented on source/image-handler/image-request.js:
```diff
@@ -137,8 +137,11 @@ class ImageRequest {
             // Parse the key from the end of the path
-            const key = (event["path"]).split("/");
-            return key[key.length - 1];
+            //const key = (event["path"]).split("/");
```
Delete these two lines? or add comment?

tiagopmello commented:
Hi guys, I am not a developer, actually I am a curious retailer that managed to deploy the serverless-image-handle script on AWS. I use it a lot for all company reports on PowerBI.
I’ve the template v2 but it stopped working due to the changes on lambda environment to sharpJS I’ve heard. 
I’ve wiped my cloudformation stack on AWS and created a new one with the template v4, I knew that it would just work with images on root but I could live with that. I see though that fit-in is not working, and basically that’s almost all I use.
I´ve found rpong work fixing the template v4, but I have no idea how to implement that on my end here. Could anyone make a template with all the commits so I can deploy an cloudformation stack pointing to the updated template?
If that is not a good idea, could someone let me know how do I make the suggested commits (changes) on the code there?
Any help would be useful, I am stuck at the moment. 
Regards,
Tiago

rpong commented:
> Hi guys, I am not a developer, actually I am a curious retailer that managed to deploy the serverless-image-handle script on AWS. I use it a lot for all company reports on PowerBI.
> I’ve the template v2 but it stopped working due to the changes on lambda environment to sharpJS I’ve heard.
> I’ve wiped my cloudformation stack on AWS and created a new one with the template v4, I knew that it would just work with images on root but I could live with that. I see though that fit-in is not working, and basically that’s almost all I use.
> I´ve found rpong work fixing the template v4, but I have no idea how to implement that on my end here. Could anyone make a template with all the commits so I can deploy an cloudformation stack pointing to the updated template?
> If that is not a good idea, could someone let me know how do I make the suggested commits (changes) on the code there?
> Any help would be useful, I am stuck at the moment.
> Regards,
> Tiago

You may pack the fork by running the following:

```
#clone the repo
cd ~
git clone git@github.com:innovationlove/serverless-image-handler.git
cd serverless-image-handler/source/image-handler

#build npm depenedencies
npm run build:init && npm install --arch=x64 --platform=linux --production && npm run build:zip && npm run build:dist
```

The output of the above will be also be a zip file as follows:

`~/serverless-image-handler/source/image-handler/dist/image-handler.zip`

Go to your Serverless Image Handler Lambda function -> Scroll down to function code ->  Select Upload a .zip file -> Click Upload ->  Select the generated zip file above -> Click Save -> Click Actions and Select Publish New version.

FOLLOW AT YOUR OWN RISK.

NOTE: If you run into any issues during build, you will need Node Version 11, depending on your platform, a little googling should nudge you to the right path regarding that one.. Goodluck.

tiagopmello commented:
> > Hi guys, I am not a developer, actually I am a curious retailer that managed to deploy the serverless-image-handle script on AWS. I use it a lot for all company reports on PowerBI.
> > I’ve the template v2 but it stopped working due to the changes on lambda environment to sharpJS I’ve heard.
> > I’ve wiped my cloudformation stack on AWS and created a new one with the template v4, I knew that it would just work with images on root but I could live with that. I see though that fit-in is not working, and basically that’s almost all I use.
> > I´ve found rpong work fixing the template v4, but I have no idea how to implement that on my end here. Could anyone make a template with all the commits so I can deploy an cloudformation stack pointing to the updated template?
> > If that is not a good idea, could someone let me know how do I make the suggested commits (changes) on the code there?
> > Any help would be useful, I am stuck at the moment.
> > Regards,
> > Tiago
> 
> You may pack the fork by running the following:
> 
> ```
> #clone the repo
> cd ~
> git clone git@github.com:innovationlove/serverless-image-handler.git
> cd serverless-image-handler/source/image-handler
> 
> #build npm depenedencies
> npm run build:init && npm install --arch=x64 --platform=linux --production && npm run build:zip && npm run build:dist
> ```
> 
> The output of the above will be also be a zip file as follows:
> 
> `~/serverless-image-handler/source/image-handler/dist/image-handler.zip`
> 
> In AWS Console: Go to your Serverless Image Handler Lambda function -> Scroll down to function code -> Select Upload a .zip file -> Click Upload -> Select the generated zip file above -> Click Save -> Click Actions and Select Publish New version.
> 
> FOLLOW AT YOUR OWN RISK.
> 
> NOTE: If you run into any issues during build, you will need Node Version 11, depending on your platform, a little googling should nudge you to the right path regarding that one.. Goodluck.


Rpong, 
Thanks for the quick reply and also all the work on fixing the serveless image handler template.
I have one linux server here on work with git installed running as root (so should have all permissions needed) but the git clone did not work.

# git clone git@github.com:innovationlove/serverless-image-handler.git
Cloning into 'serverless-image-handler'...
Permission denied (publickey).
fatal: Could not read from remote repository.

Would you happen to know how to overcome this error?
Regards,
Tiago

tiagopmello commented:
Rpong, 
Thanks for the detailed information could not have done it without you. Had problems to build had to install Node and NPM then update Zlib. I’ve built this script using Debian as system node: v10.15.2 and NPM 6.10.2. When I’ve uploaded the command on AWS Lambda I’ve changed the Runtime from Node.JS 8.1 to Node.js 10.x (was supposed to do this)?
The problem is that my files are on 'sa-east-1' and when I´ve uploaded the .zip file after compiling I’ve got this error: Error occurred while GetObject. S3 Error Code: AuthorizationHeaderMalformed. S3 Error Message: The authorization header is malformed; the region 'us-east-1' is wrong; expecting 'sa-east-1'
I guess I need to change somewhere on the zip file that this code is for 'sa-east-1' do you know where ? Is there another thing that I need to change in order to build this code on a sa-east-1 s3?
Regards,
Tiago

tiagopmello commented:
\

> > Hi guys, I am not a developer, actually I am a curious retailer that managed to deploy the serverless-image-handle script on AWS. I use it a lot for all company reports on PowerBI.
> > I’ve the template v2 but it stopped working due to the changes on lambda environment to sharpJS I’ve heard.
> > I’ve wiped my cloudformation stack on AWS and created a new one with the template v4, I knew that it would just work with images on root but I could live with that. I see though that fit-in is not working, and basically that’s almost all I use.
> > I´ve found rpong work fixing the template v4, but I have no idea how to implement that on my end here. Could anyone make a template with all the commits so I can deploy an cloudformation stack pointing to the updated template?
> > If that is not a good idea, could someone let me know how do I make the suggested commits (changes) on the code there?
> > Any help would be useful, I am stuck at the moment.
> > Regards,
> > Tiago
> 
> You may pack the fork by running the following:
> 
> ```
> #clone the repo
> cd ~
> git clone git@github.com:innovationlove/serverless-image-handler.git
> cd serverless-image-handler/source/image-handler
> 
> #build npm depenedencies
> npm run build:init && npm install --arch=x64 --platform=linux --production && npm run build:zip && npm run build:dist
> ```
> 
> The output of the above will be also be a zip file as follows:
> 
> `~/serverless-image-handler/source/image-handler/dist/image-handler.zip`
> 
> In AWS Console: Go to your Serverless Image Handler Lambda function -> Scroll down to function code -> Select Upload a .zip file -> Click Upload -> Select the generated zip file above -> Click Save -> Click Actions and Select Publish New version.
> 
> FOLLOW AT YOUR OWN RISK.
> 
> NOTE: If you run into any issues during build, you will need Node Version 11, depending on your platform, a little googling should nudge you to the right path regarding that one.. Goodluck.

Rpong, 

I was able to fix the problems mentioned above. Found out that I indeed need to use Runtime from Node.JS 8.1
Added this configs under Lambda Environment Variables
REKOGNITION_REGION: us-east-1
TC_AWS_REGION: sa-east-1

And I was able to run the .zip file that I’ve compiled, but it seems to be the same as before, fit-in isn’t working and subfolders either. Is there a way to know if I compiled the file with all the changes needed? 

**I tried to use your git clone command, it didn’t work, so I’ve used like this: 

git clone https://github.com/innovationlove/serverless-image-handler**

**is this git-clone url correct?**

Since subfolders aren’t working even with this new compile I’ve put one test photo on bucket root. As you can see, fit-in didn’t work as it supposed to.
Example of output: https://d17wgxzut1imxc.cloudfront.net/fit-in/600x600/190705003.jpg
Thanks for all the help so far, I am trying to do my best to sort this out.
Regards,
Tiago

rpong commented:
hi @tiagopmello , make sure that you did clicked "Save" and `Actions -> Publish New Version`. You may also need to invalidate your whole cache to make sure the image you are trying to request is not cached by Cloudfront.  You can do this by going to your Cloudfront Distribution -> Invalidations tab, and creating a new invalidation for all your objects  by using this path`/*` (do at your own risk).

tiagopmello commented:
> hi @tiagopmello , make sure that you did clicked "Save" and `Actions -> Publish New Version` after you uploaded the zip file in the Function Code area. You may also need to invalidate your whole cache to make sure the images you are trying to request are not not cached by Cloudfront. You can do this by going to your Cloudfront Distribution -> Invalidations tab, and creating a new invalidation for all your objects by using this path`/*` (do at your own risk).

Thanks for the reply again. **Yes it was published. I`ve clicked publish new version and save**. and also I've now invalidated the whole cloudfront distribution cache, **it still didn't work**. 

**Can you confirm if the way I've closed the github fork is correct ?** 

git clone https://github.com/innovationlove/serverless-image-handler

you suggested me to use git clone git@github.com:innovationlove/serverless-image-handler.git

**But this command did not work for me it gave permissions error.**

rpong commented:
hi @tiagopmello , please try to find me "Arpee Ong" on any chat/social media channel, and i'll send you the actual build we are using.

rpong commented:
glad that the beautiful lady in the photo is now whole @tiagopmello 

https://d17wgxzut1imxc.cloudfront.net/fit-in/605x605/190705003.jpg  :)

tiagopmello commented:
> glad that the beautiful lady in the photo is now whole @tiagopmello
> 
> https://d17wgxzut1imxc.cloudfront.net/fit-in/605x605/190705003.jpg :)

Yes, you are the best. Its working flawlessly.

Thank you!!

rpong commented:
Apparently there are reports of some users not able to package and apply the function correctly, so I bundled a release ready to be used.

You may start using this PR/Fork by using this Exported Function zip: https://github.com/innovationlove/serverless-image-handler/releases/tag/4.0.1-ilove

Just replace the code that your Serverless Image Handler Function is using with the package above. USE AT YOUR OWN RISK.

carlmathisen commented:
Thank you @rpong, I really appreciate it! Hopefully your fix will make it into the main repo soon.

hayesry commented:
@rpong Thanks for your massive contribution here. I've added your changes to our review board and hope to have them merged and published shortly in the next minor release. Awesome stuff!

beyerz commented:
This PR resolves some major BC issues for me. Is there an expected time that we can see this merged back into the main repo?

carlmathisen commented:
Any news on this PR making it in?

andyfurniss4 commented:
I've tried implementing this manually by exporting the existing Lambda function code, making the changes to the files and then re-uploading a newly zipped version of the function but I'm getting the following error:

```
module initialization error: Error
at Object.hasVendoredLibvips (/var/task/node_modules/sharp/lib/libvips.js:61:13)
at Object.<anonymous> (/var/task/node_modules/sharp/lib/constructor.js:9:22)
at Module._compile (module.js:652:30)
at Object.Module._extensions..js (module.js:663:10)
at Module.load (module.js:565:32)
at tryModuleLoad (module.js:505:12)
at Function.Module._load (module.js:497:3)
at Module.require (module.js:596:17)
at require (internal/module.js:11:18)
at Object.<anonymous> (/var/task/node_modules/sharp/lib/index.js:3:15)
```

Any ideas?

rpong commented:
> I've tried implementing this manually by exporting the existing Lambda function code, making the changes to the files and then re-uploading a newly zipped version of the function but I'm getting the following error:
> 
> ```
> module initialization error: Error
> at Object.hasVendoredLibvips (/var/task/node_modules/sharp/lib/libvips.js:61:13)
> at Object.<anonymous> (/var/task/node_modules/sharp/lib/constructor.js:9:22)
> at Module._compile (module.js:652:30)
> at Object.Module._extensions..js (module.js:663:10)
> at Module.load (module.js:565:32)
> at tryModuleLoad (module.js:505:12)
> at Function.Module._load (module.js:497:3)
> at Module.require (module.js:596:17)
> at require (internal/module.js:11:18)
> at Object.<anonymous> (/var/task/node_modules/sharp/lib/index.js:3:15)
> ```
> 
> Any ideas?

easier to just use this pre-built one: https://github.com/awslabs/serverless-image-handler/pull/130#issuecomment-517566608 goodluck and let us know.

GeorgeBellTMH commented:
Any updates?

komandar commented:
> > I've tried implementing this manually by exporting the existing Lambda function code, making the changes to the files and then re-uploading a newly zipped version of the function but I'm getting the following error:
> > ```
> > module initialization error: Error
> > at Object.hasVendoredLibvips (/var/task/node_modules/sharp/lib/libvips.js:61:13)
> > at Object.<anonymous> (/var/task/node_modules/sharp/lib/constructor.js:9:22)
> > at Module._compile (module.js:652:30)
> > at Object.Module._extensions..js (module.js:663:10)
> > at Module.load (module.js:565:32)
> > at tryModuleLoad (module.js:505:12)
> > at Function.Module._load (module.js:497:3)
> > at Module.require (module.js:596:17)
> > at require (internal/module.js:11:18)
> > at Object.<anonymous> (/var/task/node_modules/sharp/lib/index.js:3:15)
> > ```
> > 
> > 
> > Any ideas?
> 
> easier to just use this pre-built one: [#130 (comment)](https://github.com/awslabs/serverless-image-handler/pull/130#issuecomment-517566608) goodluck and let us know.

Worked very well for me in combination with #157

padre commented:
@rpong Thanks for your contribution.

But there is a problem: if the filename contains square brackets, the regular expression does not work.

Example: [test].jpg does not work (throw error 400)

rpong commented:
> > > I've tried implementing this manually by exporting the existing Lambda function code, making the changes to the files and then re-uploading a newly zipped version of the function but I'm getting the following error:
> > > ```
> > > module initialization error: Error
> > > at Object.hasVendoredLibvips (/var/task/node_modules/sharp/lib/libvips.js:61:13)
> > > at Object.<anonymous> (/var/task/node_modules/sharp/lib/constructor.js:9:22)
> > > at Module._compile (module.js:652:30)
> > > at Object.Module._extensions..js (module.js:663:10)
> > > at Module.load (module.js:565:32)
> > > at tryModuleLoad (module.js:505:12)
> > > at Function.Module._load (module.js:497:3)
> > > at Module.require (module.js:596:17)
> > > at require (internal/module.js:11:18)
> > > at Object.<anonymous> (/var/task/node_modules/sharp/lib/index.js:3:15)
> > > ```
> > > 
> > > 
> > > Any ideas?
> > 
> > 
> > easier to just use this pre-built one: [#130 (comment)](https://github.com/awslabs/serverless-image-handler/pull/130#issuecomment-517566608) goodluck and let us know.
> 
> Worked very well for me in combination with #157

Glad to help, btw just released the same package we use with Node 12.x -> https://github.com/innovationlove/serverless-image-handler/releases/tag/4.0.3

Goodluck.

rpong commented:
> @rpong Thanks for your contribution.
> 
> But there is a problem: if the filename contains square brackets, the regular expression does not work.
> 
> Example: [test].jpg does not work (throw error 400)

Will look into this within the week.

padre commented:
@rpong Thank you very much, again! But I don't think it's working properly. How do you crop an image to exactly a specified size?

Before upgrading, if I didn't use the fit-in argument, the image auto-cropped and auto-resized to be EXACTLY the specified size. Now domain.com/fit-in/50x50/image.jpg and domain.com/50x50/image.jpg  produces the same result :-(

Vita1ik commented:
Why in this case `
https://cloudfront-fake-endpoint.net/fit-in/400x0/image_id.jpg` image is not covered, I receive an empty object(

It doesn't work with lambda which runs on node.js 12.x, but I have the older Serverless Image Handler that runs the lambda on node.js 8.10 and this one works properly

OBrown92 commented:
Got also some problems with the newest version 4.1. Is it still possible to deploy version 4 and then upgrade the node version with the release from innovationlove?

Vita1ik commented:
> Why in this case ` https://cloudfront-fake-endpoint.net/fit-in/400x0/image_id.jpg` image is not covered, I receive an empty object(
> 
> It doesn't work with lambda which runs on node.js 12.x, but I have the older Serverless Image Handler that runs the lambda on node.js 8.10 and this one works properly

`filters:upscale()` solved my issue =)

beomseoklee commented:
Thanks for your effort here, @rpong 
We have updated our solution, and your pull request has been merged into our master branch. However, as we've merged many pull requests at this time, the source code might not exactly same as what you've done. In addition, due to our internal process, we've merged source code manually, so please understand how your effort has been merged.

You can refer to the recent changes [here](https://github.com/awslabs/serverless-image-handler/blob/master/CHANGELOG.md)

ijhar8 commented:
@rpong this solution is not working for image size more thane 4MB .
getting error .
{"message": "Internal server error"}

OBrown92 commented:
I think this is more a Aws Lambda problem because the request and response size is limited: 
`Request and response body payload size are maximized to 6 MB.` ([aws](https://docs.aws.amazon.com/lambda/latest/dg/gettingstarted-limits.html))

giovannipds commented:
Is there a way to use the old `fit-in` behavior url without having to overwrite the code?

cxrod commented:
@giovannipds the old fit-it is working on the last release. Just update your stack with the latest template.

giovannipds commented:
@cxrod seriously? I probably have edited the code before putting the stack up the last time I've updated here (29 days ago).

giovannipds commented:
@cxrod I won't re-do this right now since it's already working as I need on my stack 😆 but thanks for letting me know.

cxrod commented:
@giovannipds sorry for the late reply! I was observing for a long time the PR with the fix, until it was released on February

giovannipds commented:
@cxrod no worries! o/
